File: Linked_List/08-Start_Deletion/start_deletion.py

class Node:
    def __init__(self, data, next=None):
        self.data = data
        self.next = next

def tarverse_linked_list( head: Node ):
    """Traverse and print the data of the linked list."""
    if not head: print("Linked List is Empty.")
    else:
        mover = head
        while mover:
            print(mover.data, end=" -> " if mover.next else "\n")
            mover = mover.next

'''
PROBLEM - Given a Head node, delete the first node and return the new head.

APPROACH
---------
LOGIC - 
In python, the not used space is freed automatically, so we do not have to free that manually
We will just move the head to it's next pointer
return new head

TIME COMPLEXITY - O(1) 
SPACE COMPLEXITY - O(1)
'''

def start_deletion( head: Node ):
    if not head:
        print("Linked List is Empty!")
        return None
    
    head = head.next 
    return head

File: Linked_List/08-Start_Deletion/start_deletion.cpp
#include "../Linked_List.h"
#include <iostream>
using namespace std;

/*
PROBLEM - Given a Head node, delete the first node and return the new head.

APPROACH
---------
LOGIC - 
We will save the head in another varaible, cause if we directly move the head forward, then there will be memory leak
After saving, then move the head forward
Free the space of the previous head
return new head

TIME COMPLEXITY - O(1) 
SPACE COMPLEXITY - O(1)
*/

Node *start_deletion( Node *head ){
    if (!head) {
        cout << "Linked List is Empty!" << endl;
        return nullptr;
    }
    Node *to_free = head; // Save the head in another pointer variable
    head = head -> next;  // make the head it's next
    delete to_free;       // delete or free the node
    return head;
}
